providers/implementations/ciphers/cipher_chacha20_hw.c

```diff
@@ -102,6 +92,11 @@ static int chacha20_cipher(PROV_CIPHER_CTX *bctx, unsigned char *out,
         memset(ctx->buf, 0, sizeof(ctx->buf));
         ChaCha20_ctr32(ctx->buf, ctx->buf, CHACHA_BLK_SIZE,
                        ctx->key.d, ctx->counter);
+
+        /* propagate counter overflow */
+        if (!++ctx->counter[0])
+            ctx->counter[1]++;
+            
         for (n = 0; n < rem; n++)
             out[n] = in[n] ^ ctx->buf[n];
         ctx->partial_len = rem;
```

Comment:
@t8m that `partial_len` bookkeeping is already happening here.

Reply:
But that is called only if `rem > 0` here. However you're right that it does not make a difference because in the next call to chacha20_cipher() `partial_len == CHACHA_BLK_SIZE` is equivalent to `partial_len == 0`. But I would still prefer keeping the reset.

Reply:
> But I would still prefer keeping the reset.

I personally find that strange because, to a crypto developer, in any stream cipher `ctx->partial_len = 0` generally indicates that there are keystream bytes actually available in the buffer, but at the point in the code where you want the reset, there in fact is not?